pages/en/zkapps/how-to-write-a-zkapp.mdx

```diff
@@ -207,121 +205,192 @@ For a full list, see the <DocLink copy="SnarkyJS reference" url="/zkapps/snarkyj
+The `increment()` method fetches the current on-chain state `x` with `this.x.get()`.
+Later, it sets the new state to `x + 1` using `this.x.set()`. Simple!
+
+There's another line though, which looks weird at first:
+
+```ts
+this.x.assertEquals(x);
+```
+
+To understand it, we have to take a step back, and understand what it means to "use an on-chain value" during off-chain execution.
+
+For sure, when we use an on-chain value, we have to _prove_ that this is the on-chain value. Verification has to fail if it's a different value! Otherwise, a malicious user could modify SnarkyJS and make it just use any other value than the current on-chain state -- breaking our zkApp.
+
+To prevent that, we link "`x` at proving time" to be the same as "`x` at verification time". We call this a _precondition_ -- a condition that is checked by the verifier (a Mina node) when it receives the proof in a transaction. This is what `this.x.assertEquals(x)` does: it adds the precondition that `this.x` -- the on-chain state at verification time -- has to equal `x` -- the value we fetched from the chain on the client-side. In zkSNARK language, `x` becomes part of the public input.
+
+Side note: `this.<state>.assertEquals` is more flexible than equating with the current value. For example, `this.x.assertEquals(10)` fixes the on-chain `x` to the number `10`.
+
+<Alert kind="note">
 
-The method then sets the new state to `x + 1` using `this.x.set()`.
+Why didn't we just make `this.x.get()` add the precondition, automatically, so that you didn't have to write `this.x.assertEquals(x)`?
+Well, we like to keep things explicit. The assertion reminds us that we add logic which can make the proof fail: If `x` isn't the same at verification time, the transaction will be rejected. So, reading on-chain values has to be done with care if many users are supposed to read and update state concurrently. It is applicable in some situations, but might cause races, and call for workarounds, in other situations.
+One such workaround is the use of actions -- see <DocLink copy="Actions and Reducer" url="/zkapps/advanced-concepts/#actions-and-reducer" />
 
-The methods `.get()` and `.set()` are defined on every state property.
+</Alert>
 
 ##### Assertions
 
+<!-- TODO: this is _slightly_ misplaced now, because I already had assertEquals earlier -->
+
 Let's modify the `increment()` method to accept a parameter:
 
 ```ts
 class HelloWorld extends SmartContract {
   @state(Field) x = State<Field>();
 
-  // ...
+  @method increment(xPlus1: Field) {
+    const x = this.x.get();
+    this.x.assertEquals(x);
 
-  @method async increment(xPlus1: Field) {
-    const x = await this.x.get();
     x.add(1).assertEquals(xPlus1);
+
     this.x.set(xPlus1);
   }
 }
 ```
 
-Here, after obtaining the current state `x`, we make an _assertion_: `x.add(1).assertEquals(xPlus1);`.
+Here, after obtaining the current state `x` and fixing it to the on-chain value, we make another assertion:
```

Comment:
Could we phrase this `and fixing it to the on-chain value` differently? It's reads unclear

Maybe something around "enuring that x is this value when our transaction is processed by the Mina network otherwise it will be rejected"? 

wdyt? Not strong feelings. Just feels unclear

Reply:
went with "asserting that it equals the on-chain value"